### ISSUE TITLE
`recovery_test_bft` failing in CI

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#2853]
The `recovery_test_bft` test currently has a very low pass rate, and is failing most Daily CI attempts:

![image](https://user-images.githubusercontent.com/6000239/127514371-2c6ab91d-8e30-463b-a4c7-78eb5d3f55fa.png)

The failure appears to be a total failure to connect, for a newly launched node. I don't yet have any more debugging information, but suspect this may be channels-related again.

**Comment by eddyashton** [comment]
Representative logs:

https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=29884&view=logs&j=383d248c-4494-5797-d98f-6cef5140601e&t=d3511c0a-aced-5990-4c7d-8e47aebb8200&l=32383

**Comment by eddyashton** [comment]
The node attempting to join is stuck in a loop of accepting then immediately receiving an error on new SSL connections:

```
2021-07-29T14:55:30.894189Z -0.010 0   [debug] ../src/enclave/rpc_sessions.h:204    | Closing a session inside the enclave: 293
2021-07-29T14:55:30.894410Z -0.002 0   [trace] ../src/enclave/tls_endpoint.h:455    | TLS 293 on handshake: SSL - A fatal alert message was received from our peer
2021-07-29T14:55:30.894477Z        100 [debug] ../src/host/rpc_connections.h:237    | rpc stop from enclave 293, Error
2021-07-29T14:55:30.894578Z -0.002 0   [trace] ../src/http/http_endpoint.h:48       | recv called with 7 bytes
2021-07-29T14:55:30.894634Z -0.002 0   [trace] ../src/enclave/tls_endpoint.h:120    | Requesting up to 4096 bytes
2021-07-29T14:55:30.894692Z        100 [debug] ../src/host/rpc_connections.h:245    | rpc closed from enclave 293
2021-07-29T14:55:30.894747Z -0.002 0   [debug] ../src/enclave/rpc_sessions.h:204    | Closing a session inside the enclave: 293
2021-07-29T14:55:30.986126Z        100 [debug] ../src/host/rpc_connections.h:86     | rpc accept 294
2021-07-29T14:55:30.986540Z        100 [debug] ../src/host/rpc_connections.h:40     | rpc read 294: 517
2021-07-29T14:55:31.015696Z -0.001 0   [debug] ../src/enclave/rpc_sessions.h:173    | Accepting a session inside the enclave: 294
2021-07-29T14:55:31.016099Z        100 [debug] ../src/host/rpc_connections.h:216    | rpc write from enclave 294: 96
2021-07-29T14:55:31.016238Z        100 [debug] ../src/host/rpc_connections.h:216    | rpc write from enclave 294: 469
2021-07-29T14:55:31.016672Z        100 [debug] ../src/host/rpc_connections.h:40     | rpc read 294: 7
2021-07-29T14:55:31.016788Z        100 [debug] ../src/host/tcp.h:777                | TCP on_read: end of file
2021-07-29T14:55:31.016871Z        100 [debug] ../src/host/rpc_connections.h:51     | rpc disconnect 294
2021-07-29T14:55:31.024742Z        100 [debug] ../src/host/rpc_connections.h:216    | rpc write from enclave 294: 254
2021-07-29T14:55:31.024823Z        100 [debug] ../src/host/tcp.h:343                | Disconnected: Ignoring write of size 254
2021-07-29T14:55:31.024884Z        100 [debug] ../src/host/rpc_connections.h:216    | rpc write from enclave 294: 28
2021-07-29T14:55:31.024933Z        100 [debug] ../src/host/tcp.h:343                | Disconnected: Ignoring write of size 28
2021-07-29T14:55:31.024999Z        100 [debug] ../src/host/rpc_connections.h:216    | rpc write from enclave 294: 9
2021-07-29T14:55:31.025045Z        100 [debug] ../src/host/tcp.h:343                | Disconnected: Ignoring write of size 9
2021-07-29T14:55:31.025106Z -0.011 0   [trace] ../src/http/http_endpoint.h:48       | recv called with 517 bytes
2021-07-29T14:55:31.025172Z -0.011 0   [trace] ../src/enclave/tls_endpoint.h:120    | Requesting up to 4096 bytes
2021-07-29T14:55:31.025230Z -0.011 0   [debug] ../src/enclave/rpc_sessions.h:204    | Closing a session inside the enclave: 294
2021-07-29T14:55:31.025466Z -0.001 0   [trace] ../src/enclave/tls_endpoint.h:455    | TLS 294 on handshake: SSL - A fatal alert message was received from our peer
2021-07-29T14:55:31.025533Z        100 [debug] ../src/host/rpc_connections.h:237    | rpc stop from enclave 294, Error
```

I don't see logs for the other end of those connection attempts on the primary, but it does appear to be so busy catching up the other node that it has a large backlog of work (and its enclave time offsets keep increasing):

```
2021-07-29T14:55:29.679352Z -8.846 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9765
2021-07-29T14:55:29.679416Z -8.846 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000005]: 152 to 156 (152)
2021-07-29T14:55:29.683081Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [15370]: 156, 151
2021-07-29T14:55:29.683463Z -8.850 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9766
2021-07-29T14:55:29.683532Z -8.850 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000005]: 157 to 157 (152)
2021-07-29T14:55:29.683820Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [915]: 157, 156
2021-07-29T14:55:29.683894Z -8.850 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9767
2021-07-29T14:55:29.683956Z -8.850 0   [trace] ../src/node/channels.h:164           | <- n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=5561
2021-07-29T14:55:29.684019Z -8.851 0   [debug] ../src/consensus/aft/raft.h:2480     | Recv append entries response to n[0000000000000000000000000000000000000000000000000000000000000004] from n[0000000000000000000000000000000000000000000000000000000000000005]: failed
2021-07-29T14:55:29.684090Z -8.851 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000005]: 146 to 151 (152)
2021-07-29T14:55:29.685246Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [4744]: 151, 145
2021-07-29T14:55:29.685416Z -8.852 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9768
2021-07-29T14:55:29.685483Z -8.852 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000005]: 152 to 156 (152)
2021-07-29T14:55:29.688898Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [15370]: 156, 151
2021-07-29T14:55:29.689397Z -8.856 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9769
2021-07-29T14:55:29.689459Z -8.856 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000005]: 157 to 157 (152)
2021-07-29T14:55:29.689722Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [915]: 157, 156
2021-07-29T14:55:29.689790Z -8.856 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=9770
2021-07-29T14:55:29.689847Z -8.856 0   [trace] ../src/node/channels.h:164           | <- n[0000000000000000000000000000000000000000000000000000000000000005]: node msg with nonce=5562
2021-07-29T14:55:29.689921Z -8.856 0   [debug] ../src/consensus/aft/raft.h:2489     | Recv append entries response to n[0000000000000000000000000000000000000000000000000000000000000004] from n[0000000000000000000000000000000000000000000000000000000000000005] for index 157: success
2021-07-29T14:55:29.689977Z -8.857 0   [debug] ../src/consensus/aft/raft.h:2924     | In update_commit, new_commit_cft_idx: 157, new_commit_bft_idx:0. last_idx: 157
2021-07-29T14:55:29.690032Z -8.857 0   [debug] ../src/consensus/aft/raft.h:2951     | Commit if possible 0 (ci: 152) (ti 0)
```

**Comment by eddyashton** [comment]
It seems the original error is the primary (node 4)'s inability to communicate with node 5, though I don't yet understand why that blocks node 6 from joining. Those communication errors start with a pattern like this:

```
2021-07-29T14:55:10.313117Z -0.017 0   [info ] ../src/node/channels.h:684           | Node channel with n[0000000000000000000000000000000000000000000000000000000000000004] is now established.
2021-07-29T14:55:10.313188Z -0.017 0   [trace] ../src/node/channels.h:690           | Node certificate serial numbers: node=449092B4406369684B371017FB2497D5 peer=BC3CB63465A98B72E2F58F20671BF41E
2021-07-29T14:55:10.313251Z -0.017 0   [trace] ../src/node/channels.h:164           | <- n[0000000000000000000000000000000000000000000000000000000000000004]: node msg with nonce=1
2021-07-29T14:55:10.313316Z -0.017 0   [trace] ../src/node/channels.h:172           | Changing to next channel receive key
2021-07-29T14:55:10.313377Z -0.017 0   [fail ] ../src/node/channels.h:807           | Failed to verify node message from n[0000000000000000000000000000000000000000000000000000000000000004]
2021-07-29T14:55:10.313439Z -0.017 0   [info ] ../src/consensus/aft/raft.h:866      | Dropped invalid message from n[0000000000000000000000000000000000000000000000000000000000000004]
2021-07-29T14:55:10.313501Z -0.017 0   [trace] ../src/node/channels.h:164           | <- n[0000000000000000000000000000000000000000000000000000000000000004]: node msg with nonce=2
2021-07-29T14:55:10.313563Z -0.017 0   [fail ] ../src/node/channels.h:807           | Failed to verify node message from n[0000000000000000000000000000000000000000000000000000000000000004]
2021-07-29T14:55:10.313624Z -0.017 0   [info ] ../src/consensus/aft/raft.h:866      | Dropped invalid message from n[0000000000000000000000000000000000000000000000000000000000000004]
2021-07-29T14:55:10.313686Z -0.017 0   [trace] ../src/node/channels.h:164           | <- n[0000000000000000000000000000000000000000000000000000000000000004]: node msg with nonce=3
```

Node 5 has completed a key exchange, receives a `nonce=1` message indicating use of the new key, but is then unable to verify anything from node 4. As far as I can tell, node 5 never sends anything after this point, and yet node 4 keeps processing append entries failures from 5, and producing new append entries in response.

**Comment by eddyashton** [comment]
I've attached the node logs of 2 runs, one failing (bad) and one passing (good). The interesting bits are in `4.out`, `5.out`, and perhaps `6.out`.

[bad.tar.gz](https://github.com/microsoft/CCF/files/6901800/bad.tar.gz)
[good.tar.gz](https://github.com/microsoft/CCF/files/6901801/good.tar.gz)

**Comment by eddyashton** [comment]
Note that even in the successful log, the primary still has a huge enclave clock drift (~8s!) while it catches other node's up:

```
2021-07-29T15:11:03.659404Z -8.219 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000007]: node msg with nonce=3910
2021-07-29T15:11:03.659494Z -8.219 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000007]: 108 to 112 (164)
2021-07-29T15:11:03.660839Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [4189]: 112, 107
2021-07-29T15:11:03.660977Z -8.221 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000007]: node msg with nonce=3911
2021-07-29T15:11:03.661037Z -8.221 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000007]: 113 to 117 (164)
2021-07-29T15:11:03.662162Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [3987]: 117, 112
2021-07-29T15:11:03.662324Z -8.222 0   [trace] ../src/node/channels.h:785           | -> n[0000000000000000000000000000000000000000000000000000000000000007]: node msg with nonce=3912
2021-07-29T15:11:03.662384Z -8.222 0   [debug] ../src/consensus/aft/raft.h:1303     | Send append entries from n[0000000000000000000000000000000000000000000000000000000000000004] to n[0000000000000000000000000000000000000000000000000000000000000007]: 118 to 122 (164)
2021-07-29T15:11:03.663938Z        100 [debug] ../src/host/node_connections.h:321   | send AE to node 0000000000 [4285]: 122, 117
```

I think this is part of the [inefficient catchup logic](https://github.com/microsoft/CCF/issues/1451) - every time we get a single negative response, we try to send the entire ledger history again. If this disagrees for something foundational, early, we're essentially running an amplification attack on ourselves (one catchup attempt sends 5 messages, which returns 5 NACKs, each of which produces 5 catchup messages for 25 catchup messages, which return 25 NACKs, which...).

Still not clear how this produces a total failure, since it appears to be more severe than simply a timeout, but this is something we should also fix (and makes debugging these real bugs much harder).

**Comment by eddyashton** [comment]
Trimming to specific log lines, its clear that in a successful run we don't actually reach key rotation, whereas we do attempt a channel key rotation in the failing test:

```
54:55.376 |05|   Node channel with n[0000000000000000000000000000000000000000000000000000000000000004] is now established. (channels.h:684)
54:55.376 |05|   Node certificate serial numbers: node=449092B4406369684B371017FB2497D5 peer=BC3CB63465A98B72E2F58F20671BF41E (channels.h:690)
54:55.376 |05|   <- n[0000000000000000000000000000000000000000000000000000000000000004]: node msg with nonce=4631 (channels.h:164)
54:55.376 |05|   Reached message limit (1+4631), triggering new key exchange (channels.h:206)
54:55.376 |05|   Initiating node channel with n[0000000000000000000000000000000000000000000000000000000000000004]. (channels.h:707)

54:55.382 |05|   key_exchange_init -> n[0000000000000000000000000000000000000000000000000000000000000004] node serial: 449092B4406369684B371017FB2497D5 (channels.h:742)

54:55.383 |05|   -> n[0000000000000000000000000000000000000000000000000000000000000004]: node msg with nonce=1 (channels.h:785)
```

And I've totally lost track of what I was looking at here. A problem for tomorrow.

**Comment by eddyashton** [comment]
So it looks like the root cause of this stems from the `commit_idx`, and specifically my change in #2835. If the primary's `commit_idx` is in their new term (because the old nodes helped them advance it before they died), then that primary is never able to catch up the new nodes and we get this failure. If however they still have an _old_ `commit_idx`, then they are successfully able to share the end of the old term and the start of the new one. I don't understand precisely why this happens (in fact I'm worried this only works because it allows persisted data to be lost), and it really shouldn't cascade into the replay spam it currently does. But I'll add a quick workaround to at least fix the test for now, and we can revisit how these Raft catchup optimisations apply to BFT when we have time.

**Comment by eddyashton** [comment]
Resolved in #2854.
